Prevent images from appearing squished when only one dimension is set

lschuyler opened the pull request:
This PR addresses the issue of images in the editor appearing squashed when only one dimension is set.

## Problem

When you set only one of the image block's dimensions (width or height) in the editor and leave the other blank (auto), Gutenberg serializes just that single value as an inline style, like this:

```
<img 
  src="…blue-shape.png" 
  alt="blue-shape.png" 
  style="width: 64px;" 
  width="228" 
  height="228" 
/>
```

In this example, the blue-shape.png image was uploaded at 228 x 228. In the block editor UI, the width was set to 64px, and the height was left as auto. 

Because many themes (like the one we were using when we discovered this) still define a fixed CSS height (in our case, 76px), the browser will render the image at 64 x 76, making it appear distorted.

In this example, this was the existing CSS:

```
.is-style-features-page-three-column-block_img img {
  width:  76px;
  height: 76px;
}
```

## This Fix

This PR adds similar logic in two places to ensure that whenever only one dimension is set by the user, the other dimension is explicitly set to auto, to preserve the aspect ratio.

This logic is added in two places:
1.	Editor rendering (packages/block-library/src/image/image.js)
2.	Front-end save output (packages/block-library/src/image/save.js)

In each file, a small style object is built that:

- Emits width: XXpx only if the user explicitly provided a width.
- Emits height: auto when height is unset or explicitly set to "auto", blocking any theme CSS height.
- Emits height: XXpx only when the user set a numeric height.

By guaranteeing `height: auto` whenever height isn't defined, this prevents themes' CSS heights from squashing the image. A theme's default width remains in effect when the user hasn't adjusted the width in the editor.

## Handling Existing Content

A new deprecation (v9) is added to handle image blocks that were saved before this fix with only `width` set and no `height` in the inline style. When one of those blocks is next saved, it will be migrated so the inline style includes `height: auto`. Integration test fixtures are included covering parse and serialisation of this migration path.

## Testing Instructions

1. Before you put this PR in place, go to Appearance -> Editor -> Design -> Styles -> CSS and add a new style definition for the image you are about to add:

```
.image-test img {
	width: 76px;
	height: 76px;
}
```

2. Create a new page.
3. Insert an image block (select any image from your Media Library) and set the width to a number (example `64`) and leave the height blank.
4. While that image is still selected, expand "Advanced" and add the `image-test` class.
5. Notice the image dimensions. Save the page and view it (outside of the editor). Right click and select "Inspect" to see the inline style. The image will appear compressed.
6. Put this PR in place and try resetting the dimensions in the editor UI again for this image by setting the width to 50 and leaving the height unset (or set to auto). 
7. Repeat step 5. This time, the image won't appear compressed.

## Screenshots (inside the Editor)

Before the fix (64 x 76):
<img width="634" alt="Screenshot 2025-06-30 at 9 44 24 PM" src="https://github.com/user-attachments/assets/c45d672e-69a5-4315-b06b-619f07653325" />

After the fix (64 x 64):
<img width="634" alt="Screenshot 2025-06-30 at 9 45 15 PM" src="https://github.com/user-attachments/assets/3b456b5d-76df-45b4-9c82-c801e052d83a" />

lschuyler committed:
Image block: force height to auto when only width is set

When a user sets only a width (or only a height) on an image block,
theme CSS can override the other dimension and squish the image.
This change sets the unspecified dimension to `auto` via inline style
so it takes precedence over any theme stylesheet.

A v9 deprecation is added so that images saved without the explicit
`height: auto` are still recognized as valid and migrated on the next
save.

lschuyler committed:
Image block: force height to auto when only width is set

When a user sets only a width (or only a height) on an image block,
theme CSS can override the other dimension and squish the image.
This change sets the unspecified dimension to `auto` via inline style
so it takes precedence over any theme stylesheet.

A v9 deprecation is added so that images saved without the explicit
`height: auto` are still recognized as valid and migrated on the next
save.

aaronrobertshaw reviewed (commented):
Thanks for the work on this @lschuyler 👍 

This is testing pretty nicely for me. 

For others looking at the deprecation, it's worth noting that the structure in this PR is correct. As older deprecations would match on their outdated markup, they'd get resaved to acquire the `height: auto` inline style as needed. Given there is no migration of attributes required, only the new v9 deprecation is needed.

While reviewing the deprecations I think I spotted one issue with `v7.migrate()`.  With the unconditional template literal, a block with `width: 300, height: undefined` migrates to `height: "undefinedpx"`. The new current `save()` then serializes that literally as `style="height: undefinedpx"`. 

The `v6.migrate()` function handles this better.

The bug there predates this PR but this one does make it worse. So I think we should fix it up now. Additionally, it's just nice whenever adding a new version for a block to ensure every past version is migrated correctly to the latest and greatest.

One other thing, is the `isEligible` check actually needed? It's conventionally for attribute schema only migrations where the block markup is technically still valid but the attributes still need migrating. I feel `isEligible` here might be redundant and slightly misleading.

More practically: after a block is migrated and re-saved with the new format, the `height` attribute is still `undefined` (it's never stored), so `isEligible` would keep returning `true` on every subsequent load. With no `migrate()` this is a no-op, but it means the deprecation system needlessly processes the block on every parse. Worth dropping `isEligible` from v9 entirely since the save() match handles it.

lschuyler committed:
Image block: fix v7 migrate for undefined dimensions, remove v9 isEligible

v7.migrate() was unconditionally wrapping width and height in template
literals, producing "undefinedpx" when either dimension was not set.
Match the safer pattern used in v6.migrate(), which guards with
typeof ... === 'number' before appending 'px'.

Remove isEligible from v9. Since height is never stored in block
attributes when left unset, isEligible would return true on every parse
of a block with only width set, causing the deprecation system to
needlessly reprocess it. The save() match handles detection correctly
without isEligible.

lschuyler committed:
Image block: fix v7 migrate for undefined dimensions, remove v9 isEligible

v7.migrate() was unconditionally wrapping width and height in template
literals, producing "undefinedpx" when either dimension was not set.
Match the safer pattern used in v6.migrate(), which guards with
typeof ... === 'number' before appending 'px'.

Remove isEligible from v9. Since height is never stored in block
attributes when left unset, isEligible would return true on every parse
of a block with only width set, causing the deprecation system to
needlessly reprocess it. The save() match handles detection correctly
without isEligible.

aaronrobertshaw reviewed (approved):
The tweaks to the deprecations look good, thanks @lschuyler 🚀 

This PR is testing as advertised for me.

| Before | After |
|---|---|
| <img width="492" height="339" alt="Image" src="https://github.com/user-attachments/assets/35cc1985-4656-4987-88cd-45a2633f066a" /> | <img width="485" height="607" alt="Image" src="https://github.com/user-attachments/assets/61ad40f7-a3ee-470f-8011-3eb0f5dbc871" /> |

LGTM 🚢
